File: GW3/LogFileHandling/Query/Tokens/TokenCompare.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace GWLogger.Backend.DataContext.Query.Tokens
{
    internal class TokenCompare : Token
    {
        readonly static string[] keywords = new string[] { "contains", "starts", "ends" };

        public override bool CanBeUsed(QueryParser parser)
        {
            parser.SkipSpaces();
            return ((parser.PeekChar() == '=') ||
                parser.PeekChar() == '<' ||
                parser.PeekChar() == '>' ||
                (parser.PeekChar() == '!' && parser.PeekChar(1) == '=') || keywords.Contains(parser.PeekString().ToLower()));
        }

        public override Token Extract(QueryParser parser)
        {
            parser.SkipSpaces();
            if (keywords.Contains(parser.PeekString().ToLower()))
                return new TokenCompare { Value = parser.NextString().ToLower() };
            if ((parser.PeekChar() == '<' || parser.PeekChar() == '>' || parser.PeekChar() == '=') && parser.PeekChar(1) != '=')
                return new TokenCompare { Value = "" + parser.NextChar() };
            return new TokenCompare { Value = "" + parser.NextChar() + parser.NextChar() };
        }
    }
}

File: GW3/GWLogger/Frontend/Query/Tokens/TokenCompare.ts
class TokenCompare extends Token
{

    public tokenType: TokenType = TokenType.TokenCompare;

    public proposals: SuggestionInterface[] = <SuggestionInterface[]>[
        { suggestion: "!=", hint: "Not equal" },
        { suggestion: "=", hint: "Equal" },
        { suggestion: ">", hint: "Bigger than" },
        { suggestion: "<", hint: "Smaller than" },
        { suggestion: ">=", hint: "Bigger or equal than" },
        { suggestion: "<=", hint: "Smaller or equal than" },
        { suggestion: "contains", hint: "Contains" },
        { suggestion: "starts", hint: "Starts with" },
        { suggestion: "ends", hint: "Ends with" }
    ];

    static keywords: string[] = ["contains", "starts", "ends"];

    public canBeUsed(parser: QueryParser): boolean
    {
        parser.skipSpaces();
        return ((parser.peekChar() == '=') ||
            parser.peekChar() == '<' ||
            parser.peekChar() == '>' ||
            (parser.peekChar() == '!' && parser.peekChar(1) == '=') || TokenCompare.keywords.indexOf(parser.peekString().toLowerCase()) > -1);
    }

    public extract(parser: QueryParser): Token
    {
        parser.skipSpaces();
        var token = new TokenCompare()
        if (TokenCompare.keywords.indexOf(parser.peekString().toLowerCase()) > -1)
        {
            token.value = parser.nextString().toLowerCase()
            return token;
        }
        if ((parser.peekChar() == '<' || parser.peekChar() == '>' || parser.peekChar() == '=') && parser.peekChar(1) != '=')
        {
            token.value = parser.nextChar();
            return token;
        }
        token.value = parser.nextChar() + parser.nextChar();
        return token;
    }

    public getProposals(nextToken?: Token, afterNextToken?: Token): Token[]
    {
        if (typeof nextToken != 'undefined' && typeof afterNextToken != 'undefined')
        {
            switch (nextToken.tokenType)
            {
                case TokenType.TokenCompare:
                    return [new TokenString(), new TokenName(), new TokenNumber()];
                case TokenType.TokenString:
                case TokenType.TokenName:
                case TokenType.TokenNumber:
                    return [
                        new TokenAnd(),
                        new TokenGroup(),
                        new TokenLimit(),
                        new TokenOr(),
                        new TokenOrder(),
                        new TokenCloseParenthesis()
                    ];
            }
        }
        return [];
    }
}